cirq/work/sampler.py

```diff
@@ -133,6 +138,126 @@ def sample(
+                pstring_aggregate.update(pstring.items())
+            obs_pstrings.append(ops.PauliString(pstring_aggregate))
+
+        # List of observable settings with the same indexing as 'observables'.
+        obs_settings = list(observables_to_settings(obs_pstrings, qubits))
+        # Map of setting-groups to the settings they cover.
+        sampling_groups = group_settings_greedy(obs_settings)
+        sampling_params = {
+            max_setting: _get_params_for_setting(
+                setting=max_setting,
+                flips=[False] * num_qubits,
+                qubits=qubits,
+                needs_init_layer=False,
+            )
+            for max_setting in sampling_groups
+        }
+
+        # Parameterized circuit for observable measurement.
+        mod_program = _with_parameterized_layers(program, qubits, needs_init_layer=False)
+
+        input_params = list(params) if params else {}
+        num_input_param_values = len(list(study.to_resolvers(input_params)))
+        # Pairing of input sweeps with each required observable rotation.
+        sweeps = study.to_sweeps(input_params)
+        mod_sweep = study.ListSweep(sampling_params.values())
+        all_sweeps = [study.Product(sweep, mod_sweep) for sweep in sweeps]
+
+        # Results sampled from the modified circuit. Parameterization ensures
+        # that all 'z' results map directly to observables.
+        samples = self.sample(mod_program, repetitions=num_samples, params=all_sweeps)
+        results: List[List[float]] = [[0] * len(psums) for _ in range(num_input_param_values)]
+
+        for max_setting, grouped_settings in sampling_groups.items():
+            ev_params = sampling_params[max_setting]
+            # A set of num_samples integer-packed results for this setting.
+            series = [True] * len(samples)
+            for k, v in ev_params.items():
+                series = series & (samples[k] == v)
+
+            for sweep_idx, pr in enumerate(study.to_resolvers(input_params)):
+                subseries = series
+                for k in pr:
+                    subseries = subseries & (samples[k] == pr.value_of(k))
+                int_results = samples.loc[subseries, 'z']
+                for setting in grouped_settings:
+                    obs_idx = obs_settings.index(setting)
+                    psum = psums[obs_idx]
+                    ev = 0
+                    for pstr in psum:
+                        sum_val = 0
+                        for int_result in int_results:
+                            val = 1
+                            for q in pstr.qubits:
+                                bit_idx = num_qubits - qmap[q] - 1
+                                if (int_result >> bit_idx) & 1:
+                                    val *= -1
+                            sum_val += val
+                        ev += sum_val * pstr.coefficient
+                    results[sweep_idx][obs_idx] = ev / num_samples
```

Comment:
this section is very difficult to read and I suspect much slower than a vectorized numpy solution would be. Maybe factoring parts into helpfully-named helper functions or comments could go a long way?

Since you're already using some of the `cirq.work` functionality, consider using `observable_measurement_data._stats_from_measurements`

Reply:
Done. Also split off a helper function, but with `_stats_from_measurements` there's not much left to split off.